RZAssemblage/Private/RZProxyAssemblage.m

```diff
@@ -0,0 +1,162 @@
+//
+//  RZProxyAssemblage.m
+//  RZAssemblage
+//
+//  Created by Brian King on 3/19/15.
+//  Copyright (c) 2015 Raizlabs. All rights reserved.
+//
+
+#import "RZProxyAssemblage.h"
+#import "RZAssemblageChangeSet.h"
+#import "RZAssemblage+Private.h"
+#import "RZIndexPathSet.h"
+#import "RZAssemblageDefines.h"
+
+static char RZProxyKeyPathContext;
+
+@interface RZProxyAssemblage ()
+
+@property (copy, nonatomic, readonly) NSString *keypath;
+@property (copy, nonatomic, readonly) NSArray *nextKeyPaths;
+
+@end
+
+@implementation RZProxyAssemblage
+
+- (instancetype)initWithObject:(id)object keypath:(NSString *)keypath
+{
+    return [self initWithObject:object keypaths:@[keypath]];
+}
+
+- (instancetype)initWithObject:(id)object keypaths:(NSArray *)keypaths
+{
+    NSArray *remaining = [keypaths subarrayWithRange:NSMakeRange(1, keypaths.count - 1)];
+    return [self initWithObject:object observingKeyPath:[keypaths firstObject] nextKeyPaths:remaining];
+}
+
+- (instancetype)initWithObject:(id)object childKey:(NSString *)key
+{
+    return [self initWithObject:object observingKeyPath:key nextKeyPaths:nil];
+}
+
+- (instancetype)initWithObject:(id)object observingKeyPath:(NSString *)keypath nextKeyPaths:(NSArray *)keypaths
+{
+    NSParameterAssert(object);
+    NSParameterAssert(keypath);
+    // Add the observer before creating the proxy array so the backed array
+    // will trigger the observers.
+    [object addObserver:self
+             forKeyPath:keypath
+                options:NSKeyValueObservingOptionNew | NSKeyValueObservingOptionOld
+                context:&RZProxyKeyPathContext];
```

Comment:
Yikes, so the `if ( self )` below is merely a formality. If the `super` implementation chooses to set `self` to `nil` things will blow up.